api/common/wait_utils.go

```diff
@@ -42,57 +46,95 @@ import (
 
 type WaitGroup struct {
 	cond     *sync.Cond
+	closer   chan struct{}
 	isClosed bool
 	sessions uint64
 }
 
 func NewWaitGroup() *WaitGroup {
 	return &WaitGroup{
-		cond: sync.NewCond(new(sync.Mutex)),
+		cond:   sync.NewCond(new(sync.Mutex)),
+		closer: make(chan struct{}),
 	}
 }
 
-func (r *WaitGroup) AddSession() bool {
+// Closer returns a channel that is closed if
+// WaitGroup is in closing state
+func (r *WaitGroup) Closer() chan struct{} {
+	return r.closer
+}
+
+// AddSession manipulates the session counter by
+// adding or subtracting the delta value. Incrementing
+// the session counter is not possible and will set
+// return value to false if a close was initiated.
+// It's callers responsibility to make sure addition and
+// subtraction math is correct.
+func (r *WaitGroup) AddSession(delta int64) bool {
```

Comment:
i do think it would be nice to have `(r *WaitGroup) Done()` that does a -1 dec and doesn't check isClosed, the contract for AddSession is not straight forward and is effectively 2 contracts based on the positivity of the delta. in some places, we check the return value, in some places we don't. also, did not see any `AddSession(-2)`, only -1. then `Done()` has a very clear contract that it won't fail if the waiter has been closed and it won't return anything. i guess there is then simply a panic case for calling Done() one too many times.

Reply:
yeah, especially now +2 is gone.. I'll do another PR soon for this...